README
trane is a SAX-style parser for real-world, i.e. broken, HTML. It is
written in Erlang.

It aims to behave like browser HTML parsers, i.e. to turn anything
into an HTML parse tree. If the HTML is broken, this is of course
totally unpredictable, and all browsers does it differently.

The name is an homage to John Coltrane, another quality sax wielder.


ChangeLog
commit d6e92bc8028e655daad73038c52c1853ec68f88b
Author: mats cronqvist <masse@klarna.com>
Date:   Thu Oct 8 10:55:12 2015 +0200

    v1.0.1

src/trane.app.src

commit 286267335895c2eb119f6b6e091e28a1ee083b06
Author: mats cronqvist <masse@klarna.com>
Date:   Thu Oct 8 10:54:50 2015 +0200

    better test

rebar.config

commit c2c631db9f07091000ebcc5d5248941bc26c29a6
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Fri May 22 14:03:43 2015 +0200

    v1.0.0

AUTHORS
ChangeLog
src/trane.app.src

commit c29e76905e769964d5a8bbd974aaca11a1f80e69
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Fri May 22 14:03:17 2015 +0200

    encode/decode according to html4 form standard
    http://www.w3.org/TR/html4/interact/forms.html#h-17.13.4.2

src/trane_percent.erl

commit 2e898fa298451e7e279a2c7b805dfd30931fadc5
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed May 20 15:23:19 2015 +0200

    vsn 0.9.9

src/trane.app.src

commit 5bd92279407808ea79f932a09f26a9fc6c794932
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed May 20 15:18:08 2015 +0200

    much improved makefile

Makefile
bin/release.sh
src/trane.erl
src/trane_percent.erl
src/trane_poster.erl

commit 00c1b57732bd4609881698f00044167b363f2e96
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed May 20 13:09:50 2015 +0200

    add rfc3986 decoder

src/trane_percent.erl

commit c4a35b6529e435bf2000a8aac7fc853508b8fa64
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed May 20 11:34:12 2015 +0200

    introduce trane_percent to do RFC3986 encode/decode

src/trane_percent.erl

commit 50da58359c14f9b860a19a5c875feeeafe7f8b39
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed May 20 11:33:27 2015 +0200

    use eunit

.gitignore
Makefile
src/trane.erl

commit 07eb6d8a7a3b818ed27d7dd79e66ff4f8f017fef
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Sat Mar 28 16:08:25 2015 +0100

    ignore

.gitignore

commit 9f75d2ab0c61f1d6831c0110b3940cd8ebffd262
Author: mats cronqvist <masse@klarna.com>
Date:   Fri Mar 27 12:39:18 2015 +0100

    embryo to POST handler

src/trane_poster.erl

commit bef909f19f4059284fb3cce73390a8d555e49609
Author: mats cronqvist <masse@klarna.com>
Date:   Fri Mar 27 12:38:39 2015 +0100

    rebarify

ebin/trane.app
src/Makefile
src/trane.app.src
src/trane_app.erl
src/trane_sup.erl

commit 6c7cbd5ae2e3b38ddb743ce6a886511643c51f68
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Thu Mar 26 17:50:31 2015 +0100

    ignores

.gitignore

commit 4cc20f4e5c2363b862b68be454f4c7f96ea9d442
Author: mats cronqvist <masse@klarna.com>
Date:   Thu Oct 10 11:24:18 2013 +0200

    whitespace

src/trane.erl

commit 1878e94008827cd5152730167d7d5b72f2774c4c
Author: mats cronqvist <masse@klarna.com>
Date:   Sat Mar 17 20:19:15 2012 +0100

    otp 15 fix

src/trane.erl

commit 66281e794655bcd4d1b0c5f611ad216a059b034f
Author: mats cronqvist <masse@klarna.com>
Date:   Sat Mar 17 20:00:42 2012 +0100

    rebarified

Makefile
ebin/trane.app
rebar
rebar.config
src/Emakefile

commit 822c00ea5e28fea42d81f92ecca20a8cf6aea1e0
Author: Per Melin <p@greendale.se>
Date:   Mon Mar 12 15:09:17 2012 +0100

    Self-closing tags without attributes was broken
    Parsing <br/> would either crash or return {text,<<"&lt;br/>">>}
    depending on what came before it.

src/trane.erl

commit 01a6c327a65910a2e36dda7d51ac43481809754b
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Thu Nov 11 23:58:34 2010 +0100

    lost re

src/trane.erl

commit d3abe4f42ead6348347207402cd2fac8e04d1856
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 8 09:56:01 2010 +0100

    test case for <script> tag

src/trane.erl

commit 0d9b2c3788f9f76a456f80d3637d3f9efde7e0cd
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Sat Jun 12 11:47:22 2010 +0200

    returning text as string() instead of binary()

src/trane.erl

commit 0517b919e748f00f978f51db3d87c79117e5e656
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Wed Jun 9 13:20:45 2010 +0200

    speedup, wip

src/trane.erl

commit b0b94ae11767133bccbe75c5a8fdfaf40bd7f7f6
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Mon May 31 08:55:45 2010 +0200

    refactored match macros

src/trane.erl

commit b59d9da1bfc2aa9af1ccd4cbb6b91a7ebe646d52
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Mon May 31 01:47:45 2010 +0200

    representing the input string as a binary instead of a list

src/trane.erl

commit 07e028d0dca0c59d8f31819981460bbf797dae08
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Tue May 18 14:36:46 2010 +0200

    refactor

src/trane.erl

commit cb559b44c548e9e6d6a4135ae77ed5abc8f82457
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Tue May 18 09:08:36 2010 +0200

    makefiles

Emakefile
Makefile
src/Emakefile
src/Makefile
src/trane.erl

commit 733f93d84c4eec324064184ae7fe9c449a0fb2cc
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Fri May 14 19:19:18 2010 +0200

    handling spurious '>'

src/trane.erl

commit e0e336e7e6323b12f0087c35df13ffa50b78d322
Author: Torbjorn Tornkvist <tobbe@myggan.(none)>
Date:   Fri May 14 19:35:32 2010 +0800

    Added makefile support.

Emakefile
Makefile

commit 6e39fe23ec93976f456562a2757e97c34889f768
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Fri May 14 13:23:25 2010 +0200

    whitespace bug

.gitignore
src/trane.erl

commit f6d380b0fa5a15dd914a5dff15e498d4585be5a0
Author: mats cronqvist <mats.cronqvist@gmail.com>
Date:   Fri May 14 13:14:29 2010 +0200

    first commit

.gitignore
LICENSE
README
src/trane.erl


src/trane.erl
% -*- mode: erlang; erlang-indent-level: 2 -*-
%%% Created : 10 May 2010 by mats cronqvist <masse@kreditor.se>
%%% Copyright (c) 2010 Mats Cronqvist
%%
%% @doc
%% SAX parser for broken HTML
%% missing:
%% * handle attributes in end-tag (?)
%% @end

-module('trane').
-author('mats cronqvist').
-export([sax/3
         , wget_parse/1
         , wget_print/1]).

-ignore_xref(
   [sax/3,wget_parse/1,wget_print/1]).

sax(Str,Fun,Acc) ->
  parse(Str,Fun,Acc).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% parser
parse(Str,Fun,Acc) ->
  parse({Fun,Acc,[]},tokenize(Str)).

parse({Fun,Acc,Stack}=State,[{Token,T}]) ->
  case Token of
    eof -> eof(unroll(Stack,{Fun,Acc,[]}));
    _   -> parse(maybe_emit(Token,State),tokenize(T))
  end;
parse(State,[Token|Ts]) ->
  parse(maybe_emit(Token,State),Ts).

eof({_Fun,Acc,_Stack}) -> Acc.

maybe_emit(Token,State) ->
  case Token of
    {sc,Tag,Attrs}   -> emit({end_tag,Tag},emit({tag,Tag,Attrs},State));
    {open,Tag,Attrs} -> emit({tag,Tag,Attrs},push(Tag,State));
    {close,Tag}      -> emit_end_tag(Tag,State);
    {comment,Cm}     -> emit({comment,Cm},State);
    {'?',DT}         -> emit({'?',DT},State);
    {'!',DT}         -> emit({'!',DT},State);
    {text,<<>>}      -> State;
    {text,Text}      -> emit({text,Text},State)
  end.

emit_end_tag(Tag,State) ->
  try emit({end_tag,Tag},pop(Tag,maybe_unroll(Tag,State)))
  catch bogus -> State
  end.

emit(Token,{Fun,Acc,Stack}) -> {Fun,Fun(Token,Acc),Stack}.

push(Tag,{Fun,Acc,Stack}) -> {Fun,Acc,[Tag|Stack]}.

pop(Tag,{Fun,Acc,[Tag|Stack]}) -> {Fun,Acc,Stack}.

unroll(X,State) ->
  lists:foldl(fun(T,S)->emit({end_tag,T},S)end,State,X).

maybe_unroll(Tag,{Fun,Acc,Stack}) ->
  case lists:member(Tag,Stack) of
    true ->
      %% close all open tags in this tags scope
      {Hd,Tl} = lists:splitwith(fun(T)-> T=/=Tag end,Stack),
      unroll(Hd,{Fun,Acc,Tl});
    false->
      %% close tag has no open tag; drop it
      throw(bogus)
  end.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% tokenizer

%tz(<<>>,O)->lists:reverse(O);
%   (<<X:8/integer,R/binary>>,O)->tz(R,[X|O])
%end.


-define(m(X,B),<<X,B/binary>>). % match
-define(d(X,B),<<X:8/integer,B/binary>>). % decompose
-define(M(X,B),<<X,_/binary>> = B). % match
-define(D(X,B),<<X:8/integer,_/binary>> = B). % decompose

-define(ok(X),$a=<X,X=<$z;$A=<X,X=<$Z;$0=<X,X=<$9;X==$_;X==$-;X==$:).
-define(ws(X),X==$\s;X==$\r;X==$\n;X==$\t).
-define(dq(X),X==$").
-define(sq(X),X==$').

-define(ev(X), ?ws(X); X==$>; X==$=).

tokenize(Str) when is_integer(hd(Str))-> tokenize(list_to_binary(Str));
tokenize(Str) when is_binary(Str) -> tokenize(tz(nil,Str));

tokenize({TZ,Str}) ->
  tokenize(tz(TZ,Str));
tokenize({text,Str,Token}) ->
  case ff(text,Str) of
    {{tag,""},EStr,{text,Txt}} ->
      try [Token,{text,Txt}]++tokenize({{tag,""},EStr})
      catch _:_ -> tokenize({text,<<Txt/binary,"&lt;",EStr/binary>>,Token})
      end;
    {{text,Str},"",eof} ->
      [Token,{eof,{text,Str}}]
  end;
tokenize({TZ,Str,Token}) ->
  [{Token,{TZ,Str}}].

tz(nil,?m("<",Str))                       -> {{tag,""},ws(Str)};
tz(nil,?d(_,Str))                         -> {nil,Str};

tz({tag,""},?m("!--",Str))                -> {{comm,""},Str};
tz({tag,""},?m("!",Str))                  -> {{'!',""},ws(Str)};
tz({tag,""},?m("?",Str))                  -> {{que,""},ws(Str)};
tz({tag,""},?m("/",Str))                  -> {{end_tag,""},ws(Str)};
tz({tag,Tag},?m("/>",Str))                -> {text,Str,{sc,Tag,[]}};
tz({tag,Tag},?D(X,S))when ?ev(X)          -> {{attr,"",{Tag,[]}},ws(S)};
tz({tag,Tag},?d(X,Str))when ?ok(X)        -> {{tag,Tag++[X]},Str};

tz({'!',DT},?m(">",Str))                  -> {text,Str,{'!',dc(DT)}};
tz({'!',DT},?d(X,Str))                    -> {{'!',DT++[X]},Str};

tz({que,DT},?m("?>",Str))                 -> {text,Str,{'?',dc(DT)}};
tz({que,DT},?d(X,Str))                    -> {{que,DT++[X]},Str};

tz({comm,Comm},?m("-->",Str))             -> {text,Str,{comment,Comm}};
tz({comm,Comm},?d(X,Str))                 -> {{comm,Comm++[X]},Str};

tz({etag,Tag,Attrs},?m("/>",Str))         -> {text,Str,{sc,Tag,Attrs}};
tz({etag,"script",Attrs},?m(">",Str))     -> {script,Str,{open,"script",Attrs}};
tz({etag,"style",Attrs},?m(">",Str))      -> {style,Str,{open,"style",Attrs}};
tz({etag,Tag,Attrs},?m(">",Str))          -> {text,Str,{open,Tag,Attrs}};

tz({end_tag,Tag},?D(X,S))when ?ev(X)      -> {{end_tag,Tag,'>'},ws(S)};
tz({end_tag,Tag,'>'},?m(">",Str))         -> {text,Str,{close,dc(Tag)}};
tz({end_tag,Tag},?d(X,Str))               -> {{end_tag,Tag++[X]},Str};

tz({attr,"",{Tag,As}},?D(X,S))when ?ev(X) -> {{etag,dc(Tag),As},S};
tz({attr,"",{Tag,As}},?M("/>",S))         -> {{etag,dc(Tag),As},S};
tz({attr,A,{T,As}},?D(X,S))when ?ev(X)    -> {{eatt,{dc(A),T,As}},ws(S)};
tz({attr,A,{T,As}},?M("/>",S))            -> {{eatt,{dc(A),T,As}},ws(S)};
tz({attr,A,TAs},?d(X,Str))when ?ok(X)     -> {{attr,A++[X],TAs},Str};

tz({eatt,ATAs},?m("=",Str))               -> {{val,ATAs},ws(Str)};
tz({eatt,{A,T,As}},S)                     -> {{attr,"",{T,As++[{A,""}]}},ws(S)};

tz({val,ATAs},?m("'",Str))                -> {{sqval,"",ATAs},Str};%singlequoted
tz({val,ATAs},?m("\"",Str))               -> {{dqval,"",ATAs},Str};%doublequoted
tz({val,ATAs},Str)                        -> {{uqval,"",ATAs},Str};%unquoted

tz({sqval,V,{A,T,As}},?m("'",S))          -> {{attr,"",{T,As++[{A,V}]}},ws(S)};
tz({sqval,V,ATAs},?d(X,Str))              -> {{sqval,V++[X],ATAs},Str};

tz({dqval,V,{A,T,As}},?m("\"",S))         -> {{attr,"",{T,As++[{A,V}]}},ws(S)};
tz({dqval,V,ATAs},?d(X,Str))              -> {{dqval,V++[X],ATAs},Str};

tz({uqval,V,{A,T,As}},?D(X,S))when ?ev(X) -> {{attr,"",{T,As++[{A,V}]}},ws(S)};
tz({uqval,V,{A,T,As}},?M("/>",S))         -> {{attr,"",{T,As++[{A,V}]}},ws(S)};
tz({uqval,V,ATAs},?d(X,Str))              -> {{uqval,V++[X],ATAs},Str};

tz(script,Str)                            -> ff(script,Str);
tz(style,Str)                             -> ff(style,Str);
tz(text,Str)                              -> ff(text,Str);

tz(X,"")                                  -> {X,"",eof}.

ff(What,Str) -> ff(What,Str,0,Str).

ff(script,<<"</script>",Str/binary>>,N,Bin) ->
  {Scr,_} = split_binary(Bin,N),
  {{tag,""},<<"/script>",Str/binary>>,{text,Scr}};
ff(style,<<"</style>",Str/binary>>,N,Bin) ->
  {Scr,_} = split_binary(Bin,N),
  {{tag,""},<<"/style>",Str/binary>>,{text,Scr}};
ff(text,<<"<",Str/binary>>,N,Bin) ->
  {Scr,_} = split_binary(Bin,N),
  {{tag,""},ws(Str),{text,Scr}};
ff(_,<<>>,_,Str) ->
  {{text,Str},"",eof};
ff(What,<<_,Str/binary>>,N,Bin) ->
  ff(What,Str,N+1,Bin).

ws(?d(X,Str)) when ?ws(X) -> ws(Str);
ws(Str) -> Str.

dc(Str) -> string:to_lower(Str).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Parsing real pages
wget(Url) ->
  inets:start(),
  {ok, {_Rc, _Hdrs, Body}} = httpc:request(get, {Url, []}, [], []),
  Body.

wget_parse(Url) ->
  sax(wget(Url),fun(T,A)-> A++[T] end, []).

wget_print(Url) ->
  io:fwrite("~s~n",[wget(Url)]).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% TESTING

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").

t_test_() ->
  [?_assertEqual(
     t_sax("<!DOCTYPE bla><P a=b c=d>"),
     [{'!',"doctype bla"},
      {tag,"p",[{"a","b"},{"c","d"}]},
      {end_tag,"p"}]),
   ?_assertEqual(
     t_sax("<head><B>< p><p ></z></ b></x>"),
     [{tag,"head",[]},
     {tag,"b",[]},
     {tag,"p",[]},
     {tag,"p",[]},
     {end_tag,"p"},
     {end_tag,"p"},
     {end_tag,"b"},
     {end_tag,"head"}]),
   ?_assertEqual(
     t_sax("<tag catt xatt=\"\" batt>"),
    [{tag,"tag",[{"catt",""},{"xatt",""},{"batt",""}]},
     {end_tag,"tag"}]),
   ?_assertEqual(
     t_sax("<a href=/a/bc/d.e>x</a>"),
    [{tag,"a",[{"href","/a/bc/d.e"}]},
     {text,<<"x">>},
     {end_tag,"a"}]),
   ?_assertEqual(
     t_sax("<a>...<...</a>"),
    [{tag,"a",[]},
     {text,<<"...&lt;...">>},
     {end_tag,"a"}]),
   ?_assertEqual(
     t_sax("<P a=b c=d>hej<!-- te --><b>svejs</b>foo</p>grmp<br/><x x=y />"),
    [{tag,"p",[{"a","b"},{"c","d"}]},
     {text,<<"hej">>},
     {comment," te "},
     {tag,"b",[]},
     {text,<<"svejs">>},
     {end_tag,"b"},
     {text,<<"foo">>},
     {end_tag,"p"},
     {text,<<"grmp">>},
     {tag,"br",[]},
     {end_tag,"br"},
     {tag,"x",[{"x","y"}]},
     {end_tag,"x"}]),
   ?_assertEqual(
     t_sax("<script>visual+basic = rules;</script>"),
    [{tag,"script",[]},
     {text,<<"visual+basic = rules;">>},
     {end_tag,"script"}])
  ].

t_sax(Str) ->
  sax(Str,fun(T,A)-> A++[T] end,[]).

-endif.


src/trane_percent.erl
%% -*- mode: erlang; erlang-indent-level: 2 -*-
%%% Created : 19 May 2015 by masse <masse@cronqvi.st>

%% @doc
%% encode/decode percent encoded strings according to x-www-form-urlencoded
%% input/output is a string, i.e. [integer(I)] where 0 =< I =< 255.
%%   encode/1
%%      turns spaces into "+"
%%      turns all other reserved characters into "%XX" where "XX" is a hex
%%      Line breaks are represented as "\r\n" pairs (i.e., `%0D%0A')
%%   decode/1
%%      turns "+" into " "
%%      turns "%0D%0A" into "\n"
%%      turns all "%XX" into a Char, where "XX" is the hex value of Char.
%%      if "XX" is not hex, that instance of "%XX" is left untouched.
%% @end

-module('trane_percent').
-author('masse').
-export([encode/1,decode/1]).

-ignore_xref(
   [encode/1,decode/1]).

%% multipart/form-data uses MIME encoding, dafult is text/plain
%%     (http://www.w3.org/TR/html4/interact/forms.html#h-17.13.4.2)
%% application/x-www-form-urlencoded uses +
%%     (http://www.w3.org/TR/html4/interact/forms.html#h-17.13.4.1)
%% properly encoded URIs use %20
%%     (https://www.ietf.org/rfc/rfc3986.txt)

%% unreserved = ALPHA / DIGIT / "-" / "." / "_"
-define(is_unreserved(C),
        $a=<C, C=<$z;
           $A=<C, C=<$Z;
           $0=<C, C=<$9;
           C=:=$-;
           C=:=$.;
           C=:=$_).

-define(is_linebreak(C),
        C=:=$\n; C=:=$\v; C=:=$\f; C=:=$\r).

encode(Str) when is_list(Str) ->
  lists:foldr(fun enc/2,"",Str).

enc($ ,A) ->
  [$+|A];
enc(C,A) when ?is_unreserved(C) ->
  [C|A];
enc(C,A) when ?is_linebreak(C) ->
  [$%,$0,$D,$%,$0,$A|A];
enc(C,A) when 0=<C, C=<255 ->
  <<H1:4/integer,H2:4/integer>> = <<C>>,
  [$%,enchx(H1),enchx(H2)|A];
enc(C,_) ->
  error({bad_char,C}).

enchx(C) when C<10 -> C+$0;
enchx(C)           -> C-10+$A.


decode(Str) when is_list(Str) ->
  dec(Str,[]).

-define(is_cr(C), (C=:=$d orelse C=:=$D)).
-define(is_nl(C), (C=:=$a orelse C=:=$A)).
dec([],A) ->
  lists:reverse(A);
dec([$+|R],A) ->
  dec(R,[$ |A]);
dec([$%,$0,CR,$%,$0,NL|R],A) when ?is_cr(CR),?is_nl(NL) ->
  dec(R,[$\n|A]);
dec([$%,C1,C2|R],A) ->
  case dechx(C1,C2) of
    fail -> dec([C1,C2|R],[$%|A]);
    Char -> dec(R,[Char|A])
  end;
dec([C|R],A) when 0=<C, C=<255 ->
  dec(R,[C|A]);
dec(C,_) ->
  error({bad_char,C}).

dechx(C1,C2) ->
  case {dechx(C1),dechx(C2)} of
    {fail,_} -> fail;
    {_,fail} -> fail;
    {I1,I2}  -> <<I:8/integer>> = <<I1:4/integer,I2:4/integer>>, I
  end.

dechx(C) when $a=<C,C=<$f -> C-$a+10;
dechx(C) when $A=<C,C=<$F -> C-$a+10;
dechx(C) when $0=<C,C=<$9 -> C-$0;
dechx(_) -> fail.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% TESTING

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").

e_test_() ->
  [?_assertEqual("abcABC123"         ,encode("abcABC123")),
   ?_assertEqual("%0D%0A%0D%0A"      ,encode("\r\n")),
   ?_assertEqual("25%2525"           ,encode("25%25")),
   ?_assertEqual("Hatti+fnatti%21"   ,encode("Hatti fnatti!")),
   ?_assertError({bad_char,_}        ,encode([256])),
   ?_assertEqual("%2B"               ,encode("+")),
   ?_assertEqual("%26%3D%2A"         ,encode("&=*"))].

d_test_() ->
  [?_assertEqual("abcABC123"     ,decode("abcABC123")),
   ?_assertEqual("\n"            ,decode("%0D%0A")),
   ?_assertEqual("25%25"         ,decode("25%2525")),
   ?_assertEqual("Hatti fnatti!" ,decode("Hatti%20fnatti%21")),
   ?_assertEqual("Hatti fnatti!" ,decode("Hatti+fnatti%21")),
   ?_assertEqual("+"             ,decode("%2B")),
   ?_assertEqual("27% of 2B"     ,decode("27% of 2B")),
   ?_assertError({bad_char,_}    ,decode([256|"abc"])),
   ?_assertEqual("&=*"           ,decode("%26%3D%2A"))].

-endif.
